File: frontend/src/types/invoice.ts
/**
 * 发票（开票申请）相关类型定义。
 */

export type InvoiceTitleType = 'personal' | 'company'
export type InvoiceStatus = 'pending' | 'issued' | 'rejected' | 'cancelled'

/** 发票申请单 */
export interface InvoiceItem {
  id: number
  recipient_email: string
  title_type: InvoiceTitleType
  title_name: string
  tax_id?: string
  user_remark?: string
  amount: number
  status: InvoiceStatus
  invoice_number?: string
  invoice_date?: string
  invoice_amount?: number
  issued_at?: string
  reject_reason?: string
  email_sent: boolean
  created_at: string
  updated_at: string
  // 管理端附带的申请人信息
  user_id?: number
  user_email?: string
  user_name?: string
}

/** 可开票 / 申请单关联的订单（精简展示） */
export interface InvoiceableOrder {
  id: number
  order_type: string
  amount: number
  pay_amount: number
  payment_type: string
  status: string
  paid_at?: string
  created_at: string
}

/** 可开票额度明细（按消费来源核定，人民币） */
export interface InvoiceableSummary {
  available_amount: number // 可开票总额
  balance_amount: number // 其中：余额按量消费可开部分
  plan_amount: number // 其中：套餐购买可开部分
  invoiced_amount: number // 已被待开/已开申请占用（已扣除）
}

/** 提交开票申请的入参 */
export interface ApplyInvoicePayload {
  recipient_email: string
  title_type: InvoiceTitleType
  title_name: string
  tax_id?: string
  user_remark?: string
  // 申请开票金额（人民币）；省略或 <=0 表示按可开票额度全额开票
  amount?: number
}

/** 申请单详情（含关联订单） */
export interface InvoiceDetail {
  invoice: InvoiceItem
  orders: InvoiceableOrder[]
}

/** PDF 识别出的发票字段（用于开票弹窗预填） */
export interface ParsedInvoicePDF {
  invoice_number: string
  invoice_date: string | null
  invoice_amount: number | null
}

File: backend/internal/handler/dto/invoice.go
package dto

import (
	"time"

	dbent "github.com/Wei-Shaw/sub2api/ent"
)

// InvoiceItem 发票申请单响应。带 User* 的字段仅管理端填充。
type InvoiceItem struct {
	ID             int64      `json:"id"`
	RecipientEmail string     `json:"recipient_email"`
	TitleType      string     `json:"title_type"`
	TitleName      string     `json:"title_name"`
	TaxID          string     `json:"tax_id,omitempty"`
	UserRemark     string     `json:"user_remark,omitempty"`
	Amount         float64    `json:"amount"`
	Status         string     `json:"status"`
	InvoiceNumber  string     `json:"invoice_number,omitempty"`
	InvoiceDate    *time.Time `json:"invoice_date,omitempty"`
	InvoiceAmount  *float64   `json:"invoice_amount,omitempty"`
	IssuedAt       *time.Time `json:"issued_at,omitempty"`
	RejectReason   string     `json:"reject_reason,omitempty"`
	EmailSent      bool       `json:"email_sent"`
	CreatedAt      time.Time  `json:"created_at"`
	UpdatedAt      time.Time  `json:"updated_at"`

	UserID    int64  `json:"user_id,omitempty"`
	UserEmail string `json:"user_email,omitempty"`
	UserName  string `json:"user_name,omitempty"`
}

// InvoiceOrderItem 申请单关联/可开票订单的精简展示。
type InvoiceOrderItem struct {
	ID          int64      `json:"id"`
	OrderType   string     `json:"order_type"`
	Amount      float64    `json:"amount"`     // 到账金额（含赠送，仅参考）
	PayAmount   float64    `json:"pay_amount"` // 实付金额（开票口径）
	PaymentType string     `json:"payment_type"`
	Status      string     `json:"status"`
	PaidAt      *time.Time `json:"paid_at,omitempty"`
	CreatedAt   time.Time  `json:"created_at"`
}

// InvoiceFromEnt 将 ent 实体映射为响应 DTO；includeUser=true 时附带申请人信息（管理端）。
func InvoiceFromEnt(ir *dbent.InvoiceRequest, includeUser bool) InvoiceItem {
	item := InvoiceItem{
		ID:             int64(ir.ID),
		RecipientEmail: ir.RecipientEmail,
		TitleType:      ir.TitleType,
		TitleName:      ir.TitleName,
		Amount:         ir.Amount,
		Status:         ir.Status,
		InvoiceDate:    ir.InvoiceDate,
		InvoiceAmount:  ir.InvoiceAmount,
		IssuedAt:       ir.IssuedAt,
		EmailSent:      ir.EmailSent,
		CreatedAt:      ir.CreatedAt,
		UpdatedAt:      ir.UpdatedAt,
	}
	if ir.TaxID != nil {
		item.TaxID = *ir.TaxID
	}
	if ir.UserRemark != nil {
		item.UserRemark = *ir.UserRemark
	}
	if ir.InvoiceNumber != nil {
		item.InvoiceNumber = *ir.InvoiceNumber
	}
	if ir.RejectReason != nil {
		item.RejectReason = *ir.RejectReason
	}
	if includeUser {
		item.UserID = ir.UserID
		item.UserEmail = ir.UserEmail
		item.UserName = ir.UserName
	}
	return item
}

// InvoicesFromEnt 批量映射申请单。
func InvoicesFromEnt(items []*dbent.InvoiceRequest, includeUser bool) []InvoiceItem {
	out := make([]InvoiceItem, 0, len(items))
	for _, ir := range items {
		out = append(out, InvoiceFromEnt(ir, includeUser))
	}
	return out
}

// InvoiceOrdersFromEnt 批量映射订单为精简展示项。
func InvoiceOrdersFromEnt(orders []*dbent.PaymentOrder) []InvoiceOrderItem {
	out := make([]InvoiceOrderItem, 0, len(orders))
	for _, o := range orders {
		out = append(out, InvoiceOrderItem{
			ID:          int64(o.ID),
			OrderType:   o.OrderType,
			Amount:      o.Amount,
			PayAmount:   o.PayAmount,
			PaymentType: o.PaymentType,
			Status:      o.Status,
			PaidAt:      o.PaidAt,
			CreatedAt:   o.CreatedAt,
		})
	}
	return out
}
